analyses/chromosomal-instability/util/chr-break-calculate.R

```diff
@@ -185,8 +200,55 @@ break_density <- function(breaks_df = NULL,
   bins@elementMetadata@listData$total_counts <- bin_counts
 
   # Calculate and store density
-  bins@elementMetadata@listData$density <- bin_counts / window_size
-
-  # Return the GRanges object for mapping purposes
-  return(bins)
+  bins@elementMetadata@listData$density <- bin_counts / bins@ranges@width
+
+  # If specified, make unsurveyed into a GenomicRanges
+  if (!is.null(unsurveyed_bed)) {
+    # Make GRanges for uncallable ranges
+    unsurveyed_ranges <- GenomicRanges::GRanges(
+      seqnames = unsurveyed_bed$chrom,
+      ranges = IRanges::IRanges(
+        start = unsurveyed_bed$start,
+        end = unsurveyed_bed$end
+      ))
+
+    # Find the NA region(s) for each bin without combining close regions
+    na_regions <- GenomicRanges::pintersect(IRanges::findOverlapPairs(bins, unsurveyed_ranges))
+
+    # Find overlap between na_regions and bin
+    na_overlaps <- GenomicRanges::findOverlaps(bins, na_regions)
+    
+    # Get the sum of the length of all excluded regions for each bin. 
+    excluded_length_per_bin <- tapply(na_regions@ranges@width, # Get length of each sequence
+                                      na_overlaps@from, # Index of which bin it overlaps
+                                      sum) # Add up per bin
+    
+    # Get the total bin length for each bin that has excluded regions
+    bin_length <- bins[unique(na_overlaps@from)]@ranges@width
+    
+    # Calculate the percent overlap, adding up the na regions within a bin
+    pct_overlap <- excluded_length_per_bin / bin_length
+
+    # Get the bin indices that correspond to less than the cutoff
+    bin_indices <- na_overlaps@from[which(pct_overlap > perc_cutoff)]
```

Comment:
If `unique()` above on line 227 is doing something, then these pct_overlaps won't necessarily be the same indexes, right? Or is the `unique` not needed there?

Reply:
Yes. I think we found this was not exactly right. But is now fixed with most recent commit.